### PR TITLE
`MeanAveragePrecision` backend

### LINKED ISSUE #1153: `MeanAveragePrecision` (map) returns -1.0 if `max_detection_thresholds` is changed
## 🐛 Bug
I found a bug when changing the max_detection_thresholds of MeanAveragePrecision to a large number e.g. [1, 10, 1000] the result of the map calculation becomes -1.0
I wanted to increase the number of detections per image since my data has up to several thousand instances per image. I detected that the same problem occurs with the example given in the class definition.

### To Reproduce
Run the example from the function definition with max_detection_thresholds=[1, 10, 1000] instead of the default [1, 10, 100].

#### Code sample
```py
import torch
from torchmetrics.detection.mean_ap import MeanAveragePrecision
preds = [dict(
              boxes = torch.tensor([[258.0, 41.0, 606.0, 285.0]]),
              scores = torch.tensor([0.536]),
              labels = torch.tensor([0]),
             )]
target = [dict(
               boxes = torch.tensor([[214.0, 41.0, 562.0, 285.0]]),
               labels = torch.tensor([0]),
             )]
metric = MeanAveragePrecision(max_detection_thresholds=[1, 10, 1000])
metric.update(preds, target)
from pprint import pprint
pprint(metric.compute())
```
Output:
```
{'map': tensor(-1.),
 'map_50': tensor(1.),
 'map_75': tensor(1.),
 'map_large': tensor(0.6000),
 'map_medium': tensor(-1.),
 'map_per_class': tensor(-1.),
 'map_small': tensor(-1.),
 'mar_1': tensor(0.6000),
 'mar_10': tensor(0.6000),
 'mar_1000': tensor(0.6000),
 'mar_1000_per_class': tensor(-1.),
 'mar_large': tensor(0.6000),
 'mar_medium': tensor(-1.),
 'mar_small': tensor(-1.)}
```
### Expected behavior

Get the same output as with [1, 10, 100]:
```
{'map': tensor(0.6000),
 'map_50': tensor(1.),
 'map_75': tensor(1.),
 'map_large': tensor(0.6000),
 'map_medium': tensor(-1.),
 'map_per_class': tensor(-1.),
 'map_small': tensor(-1.),
 'mar_1': tensor(0.6000),
 'mar_10': tensor(0.6000),
 'mar_100': tensor(0.6000),
 'mar_100_per_class': tensor(-1.),
 'mar_large': tensor(0.6000),
 'mar_medium': tensor(-1.),
 'mar_small': tensor(-1.)}
```

### Environment
- python 3.9
- pytorch: 1.10
- TorchMetrics: 0.9.0

Thanks for your help.

### LINKED ISSUE #2035: Torchetrics MeanAveragePrecision compability with numpy >1.20.0
### To Reproduce

<details>
  <summary>Code sample used from documentation:</summary>

   ```python 


   from torch import tensor
   from torchmetrics.detection import MeanAveragePrecision

   preds = [
   dict(
        boxes=tensor([[258.0, 41.0, 606.0, 285.0]]),
        scores=tensor([0.536]),
        labels=tensor([0]),
    )
    ]

    target = [
    dict(
        boxes=tensor([[214.0, 41.0, 562.0, 285.0]]),
        labels=tensor([0]),
    )
    ]

    metric = MeanAveragePrecision(iou_type="bbox")
    metric.update(preds, target)
    from pprint import pprint
    pprint(metric.compute())
```
</details>

### Environment

- TorchMetrics version : 1.1.1
- Python & PyTorch Version (e.g., 1.0): 3.8.10 & 1.13.1+cu117
- numpy: 1.24.4
- pycocotools: https://github.com/cocodataset/cocoapi.git#subdirectory=PythonAPI

### Additional context

In [documentation](https://torchmetrics.readthedocs.io/en/latest/detection/mean_average_precision.html) for map I found "This metric utilizes the official pycocotools implementation as its backend". But there is a small confusing point: the official repository "pycocotools" and "pycocotools" package in pypi are different things. Which official pycocotools package was meant in the documentation?

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #1153
Fixes #2035
Adds a `backend` argument to `MeanAveragePrecision`. The official `pycocotools` implementation we use is flawed as there are certain corner cases were the implementation does not work. Sadly the implementation is not really maintained any more. However, it is still considered the "official" implementation and therefore we cannot do so much about this.

This PR adds the possibility of other backends for computation, starting with https://github.com/MiXaiLL76/faster_coco_eval which is a faster version and fixes at least the corner case identified in the linked issue. I still think we need to keep `pycocotools` as the default backend even if all the testing I have done shows that the two backends produce identical results. 

The goal for the future would be to reintroduce our old pytorch implementation with `backend="pytorch"`

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2034.org.readthedocs.build/en/2034/

### CODE REVIEW

File: src/torchmetrics/detection/mean_ap.py [hunk 1]
```diff
@@ -142,9 +163,16 @@ class MeanAveragePrecision(Metric):
...
+        This metric supports at the moment two different backends for the evaluation. The default backend is
+        `"pycocotools"`, which either require the official `pycocotools`_ implementation or this `fork of pycocotools`_
+        to be installed. We recommend using the fork as it is better maintained and easily avialable to install via pip:
+        `pip install pycocotools`. It is also this fork that will be install if you install `torchmetrics[detection]`.
+        The second backend is the `faster-coco-eval`_ implementation, which can be installed with
+        `pip install faster-coco-eval`. This implementation is a maintained open-source implementation that is
+        faster and corrects certain corner cases that the official implementation has. Our own testing have shown that
+        the results are identical to the official implementation. Regardless of backend we also requires you to have
+        `torchvision` version 0.8.0 or newer installed. Please install with `pip install torchvision>=0.8`
+        or `pip install torchmetrics[detection]`.
```
Comment (Borda) [inline src/torchmetrics/detection/mean_ap.py:175]:
```suggestion
        This metric supports, at the moment, two different backends for the evaluation. The default backend is
        ``"pycocotools"``, which either require the official `pycocotools`_ implementation or this `fork of pycocotools`_
        to be installed. We recommend using the fork as it is better maintained and easily available to install via pip:
        `pip install pycocotools`. It is also this fork that will be installed if you install ``torchmetrics[detection]``.
        The second backend is the `faster-coco-eval`_ implementation, which can be installed with
        ``pip install faster-coco-eval``. This implementation is a maintained open-source implementation that is
        faster and corrects certain corner cases that the official implementation has. Our own testing has shown that
        the results are identical to the official implementation. Regardless of the backend we also require you to have
        `torchvision` version 0.8.0 or newer installed. Please install with ``pip install torchvision>=0.8``
        or ``pip install torchmetrics[detection]``.
```

File: src/torchmetrics/detection/mean_ap.py [hunk 2]
```diff
@@ -189,6 +217,8 @@ class MeanAveragePrecision(Metric):
 
         average:
             Method for averaging scores over labels. Choose between "``macro``"" and "``micro``". Default is "macro"
+        backend:
+            Backend to use for the evaluation. Choose between "``pycocotools``"" and "``faster_coco_eval``".
```
Comment (Borda) [inline src/torchmetrics/detection/mean_ap.py:221]:
```suggestion
            Method for averaging scores over labels. Choose between "``"macro"`` and ``"micro"``.
        backend:
            Backend to use for the evaluation. Choose between ``"pycocotools"`` and ``"faster_coco_eval"``.
```

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
